pkg/outputs/elasticsearch/elasticsearch.go

```
@@ -0,0 +1,194 @@
+	}
+	return output.sendBulkRequests(reqs)
+}
+
+func genDocID(msg *core.Msg) string {
+	pks := []string{}
+	for _, v := range msg.DmlMsg.Pks {
+		pks = append(pks, fmt.Sprint(v))
+	}
+	return strings.Join(pks, "_")
+}
+
+func marshalError(err *elastic.ErrorDetails) string {
+	bytes, _ := json.Marshal(err)
+	return string(bytes)
+}
+
+func (output *ElasticsearchOutput) sendBulkRequests(reqs []elastic.BulkableRequest) error {
+	if len(reqs) == 0 {
+		return nil
+	}
+	bulkRequest := output.client.Bulk()
+	bulkRequest.Add(reqs...)
+	bulkResponse, err := bulkRequest.Do(context.Background())
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	for _, item := range bulkResponse.Items {
+		for action, result := range item {
+			if result.Status >= 200 && result.Status <= 299 {
+				// tags: [pipelineName, index, action(index/create/delete/update)].
+				// indices created in 6.x only allow a single-type per index, so we don't need the type as a tag.
+				metrics.OutputCounter.WithLabelValues(output.pipelineName, result.Index, action, "", "").Add(1)
+			} else if result.Status == http.StatusNotFound && action == "delete" {
+				// delete but not found, just ignore it.
+				metrics.OutputCounter.WithLabelValues(output.pipelineName, result.Index, action, "", "").Add(1)
+				continue
+			} else if result.Status == http.StatusTooManyRequests {
+				// when the server returns 429, it must be that all requests have failed.
+				time.Sleep(5 * time.Second)
+				return output.sendBulkRequests(reqs)
+			} else if output.config.IgnoreBadRequest {
+				// TODO do we need a metcris in prometheus for this?
+				log.Infof("[output_elasticsearch] Received and ignored an error from server, status: [%d], index: %s, details: %s.", result.Status, result.Index, marshalError(result.Error))
+			} else {
+				log.Panicf("[output_elasticsearch] Received an error from server, status: [%d], index: %s, details: %s.", result.Status, result.Index, marshalError(result.Error))
```

we can just return the error to scheduler with detailed error information. The scheduler will panic if multiple retries failed.